IdentityCore/src/cache/mac/MSIDMacKeychainTokenCache.m

```diff
@@ -443,44 +561,128 @@ - (MSIDCredentialCacheItem *)tokenWithKey:(__unused MSIDCacheKey *)key
+             */
+            MSIDMacCredentialStorageItem *savedItem = [self storageItemWithKey:key serializer:serializer context:context error:error];
+            MSIDMacCredentialStorageItem *currentItem = key.isShared ? self.sharedStorageItem : self.appStorageItem;
+            
+            if (key.isShared)
             {
-                [tokenList addObject:cacheItem];
+                [currentItem mergeStorageItem:savedItem];
             }
-            else
+            else if (![[currentItem storedCredentialsForKey:(MSIDDefaultCredentialCacheKey *)key] count])
             {
-                MSID_LOG_WITH_CTX(MSIDLogLevelWarning,context, @"Failed to deserialize account");
+                [currentItem mergeStorageItem:savedItem];
             }
+            
+            tokenList = [currentItem storedCredentialsForKey:(MSIDDefaultCredentialCacheKey *)key];
         }
     }
+    
     return tokenList;
 }
 
 // Remove one or more credentials from the keychain that match the key (see credentialItem:matchesKey).
-- (BOOL)removeTokensWithKey:(__unused MSIDCacheKey *)key
-                    context:(__unused id<MSIDRequestContext>)context
-                      error:(__unused NSError **)error
+- (BOOL)removeTokensWithKey:(MSIDCacheKey *)key context:(id<MSIDRequestContext>)context error:(NSError *__autoreleasing *)error
 {
-    return [self removeItemsWithKey:key context:context error:error];
+    MSID_TRACE;
+    
+    if (!key || !(key.account || key.service))
+    {
+        MSID_LOG_WITH_CTX(MSIDLogLevelError, context, @"Key is nil or one of the key attributes account or service is nil.");
+        [self createError:@"Key is nil or one of the key attributes account or service is nil."
+                   domain:MSIDErrorDomain errorCode:MSIDErrorInvalidDeveloperParameter error:error context:context];
+        return NO;
+    }
+    
+    MSIDMacCredentialStorageItem *storageItem = [self storageItemWithKey:key serializer:self.serializer context:context error:error];
+    
+    if (storageItem && [key isKindOfClass:[MSIDDefaultCredentialCacheKey class]])
+    {
+        [storageItem removeStoredCredentialForKey:(MSIDDefaultCredentialCacheKey *)key];
+        return [self saveStorageItem:storageItem key:key serializer:self.serializer context:context error:error];
+    }
+    
+    return YES;
+}
+
+- (MSIDMacCredentialStorageItem *)storageItemWithKey:(MSIDCacheKey *)key
+                                          serializer:(id<MSIDCacheItemSerializing>)serializer
+                                             context:(id<MSIDRequestContext>)context
+                                               error:(NSError **)error
+{
+    MSID_TRACE;
+    MSIDMacCredentialStorageItem *storageItem = nil;
+    NSMutableDictionary *query = [self.defaultCacheQuery mutableCopy];
+    
+    if (key.isShared)
```

Comment:
I think I've seen this code before, should we refactor into a helper method?

Reply:
Updated.